docs/device-assignment.md

```diff
@@ -0,0 +1,130 @@
+
+If `A` is assigned to `α`, we bind both `A` and `B` to vfio-pci and consider them unusable for further assignment. That means `β` cannot be started if `B` was assigned to it. We are still able to start `β` with `C`, as `C` is in separate group.
+
+## Use Cases
+
+* ML / AI
+* video processing
+* hardware / OS compatibility
+
+## KubeVirt Device Assignment
+
+To enable device assignment in KubeVirt, we need to respect both Kubernetes resource handling and virtualization's complex system requirements. To expose PCI devices to Kubelet, that further communicates them as node resources, we use a [DPI with those capabilities](https://github.com/mpolednik/linux-vfio-k8s-dpi). The DPI must be able to expose devices in uniform manner across nodes and make sure that each device that is required in assignment or affected by IOMMU is bound to vfio-pci.
+
+The VFIO endpoints, for devices requested by the pod are then exposed as host/container paths (/dev/vfio/vfio, /dev/vfio/{N, M, ...} where N, M are IOMMU group IDs) to the container in the pod.
+
+The main difficulty of this approach is reliably identifying which system device was assigned. The KubeVirt API must be verbose enough to let users request a device by it's vendor ID/product ID tuple, yet hide the actual device address. This is difficult for libvirt, as VM XML requires users to specify the device address.
+
+Because the actual device selection is done by kube-scheduler, and system may contain multiple devices with identical vendor ID/product ID that are in the same or different IOMMU groups, we need a mechanism to inspect which IOMMU group does the device belong to.
+
+To demonstrate the issue, we present an example.
+
+### Example
+
+* IOMMU group `1` contains devices
+ *	`A` (vendor ID = 1000, product ID = 1000)
+ * `B` (vendor ID = 2000, product ID = 2000)
+* IOMMU group `2` contains
+ * device `C` (vendor ID = 1000, product ID = 1000)
+* we have VM `α`, and it's corresponding pod `a`
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: α
+spec:
+  domain:
+    devices:
+      ...
+      passthrough:
+      - type: pci
+        vendor: 1000
+        device: 1000
+      ...
+    memory:
+      unit: MB
+      value: 64
+    os:
+      type:
+        os: hvm
+    type: qemu
+```
+
+In this case, KubeVirt must generate pod specification that requests device with vendor ID/product ID 1000:1000. Such specification is shown below.
+
+```yaml
+spec:
+  containers:
+  - name: a
+    ...
+    resources:
+      requests:
+              mpolednik.github.io/1000_1000: 1
+      limits:
+              mpolednik.github.io/1000_1000: 1
+```
+
+The difficult part is: how should libvirt XML generated by virt-handler look like? It is able to query the underlying system for devices and determine addresses of all devices that match vendor ID/product ID 1000:1000, but choosing a different device than the scheduler scheduled would cause the VM not to start due to IOMMU violation.
```

Comment:
What about abstracting this away?

IIUIC a IOMMUM group has  asingle "primary" device with additional supportive devices (i.e. GPU pri, supportive audio devices).

Reply:
IOMMU group can contain arbitrary devices when we talk about generic system. The idea of having a primary and supporting devices only holds on well-behaving systems (so, pretty rare).